### DISCUSSION TITLE
How CloudStack handles multiple NFS type Primary Storage

### DISCUSSION THREAD

**Discussion by shaerul** [apache/cloudstack#9073] [Q&A] [answered]
I have added a second primary storage to my CloudStack which has augmented the storage space in Dashboard.  I have gone through the CloudsStack documents and browsed through the Internet but didn't get a precise clue how ACS is managing multiple NFS storages to distribute VM image creation. I didn't find any global parameters either to control this distribution behavior.

Can someone please enlighten me?

##### ISSUE TYPE

 * Feature Idea

##### COMPONENT NAME

~~~
Multiple NFS based primary storages
~~~

##### CLOUDSTACK VERSION

~~~
4.18.1.0
~~~

##### OS / ENVIRONMENT

Ubuntu 20.04 LTS

##### SUMMARY

Wondering how the writing is managed by ACS

##### STEPS TO REPRODUCE

 * Infrastructure
 * Primary Storage
 * Add Primary Storage
 * Configuring NFS Storage
 * After Adding NFS successfully creating Instances/VMs volume

~~~

~~~

##### EXPECTED RESULTS

~~~
* Either round robin Instance Volume creation among two or multiple NFS primary storages OR
* The first one will fill up the threshold space and then ACS will start creating on the second primary storage and so on.
~~~

##### ACTUAL RESULTS

~~~
Unknown
~~~

**Comment by andrijapanicsb** [comment]
Please post your questions on the mailing list, GitHub is not the right place for it (it creates clutter) and you will get a much quicker reply on the mailing list (users@).
Thx

**Reply by yadvr** [reply]
Since this is connected to users@ ML, I don’t see why users can’t ask here. Anybody replying on users@ wouldn’t be reflected here though, but they can follow the link to advise further.

**Comment by shaerul** [comment]
Got it thanks.  Can you please delete this from here?

**Comment by yadvr** [comment]
@shaerul for all NFS primary storages, CloudStack copies the template/iso (image) on the primary storage from the secondary storage which is called preparing template on primary storage. Such a prepared or cached template is then used to deploy VMs on primary storage where for most hypervisors the VM’s boot/root disk is a copy or link to the cached image.

**Reply by shaerul** [reply]
@rohityadavcloud 
Thank you very much for all the documentation you've provided on CloudStack installation and troubleshooting. It's been incredibly helpful and has made our lives much easier.

I wanted to ask: If you have multiple NFS primary storages, how does CloudStack manage the distribution of VM volume creation across them? Does it follow round-robin or something like condensed method?

Thanks.

**Reply by yadvr** [reply, marked as answer]
@shaerul It follows a lazy deployment strategy, i.e. it copies the templates (images) to a primary storage only when it needs to create a VM's root disk on a pool. There's also a [prepareTemplate API](https://cloudstack.apache.org/api/apidocs-4.19/apis/prepareTemplate.html) that can load a template across pools to do this ahead of time.

In terms of using storage pool upon VM deployment or volume creation, it depends what deployment planning algorithm is used, pl search for these global settings (for tuning and configuration): image.store.allocation.algorithm [1], vm.allocation.algorithm [2].

[1] Image store allocation algorithm (image.store.allocation.algorithm)
firstfitleastconsumed','random' : Order in which hosts within a cluster will be considered for VM/volume allocation

[2] Vm allocation algorithm (vm.allocation.algorithm)
'random', 'firstfit', 'userdispersing', 'userconcentratedpod_random', 'userconcentratedpod_firstfit', 'firstfitleastconsumed' : Order in which hosts within a cluster will be considered for VM/volume allocation.
